bookshelf/main_test.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import re
+
+import google.auth
+import main
+import pytest
+import requests
+from retrying import retry
+from six import BytesIO
+
+
+credentials, project_id = google.auth.default()
+os.environ['GOOGLE_CLOUD_PROJECT'] = project_id
+
+
+@pytest.yield_fixture
+def app(request):
+    """This fixtures provides a Flask app instance configured for testing.
+
+    It also ensures the tests run within a request context, allowing
+    any calls to flask.request, flask.current_app, etc. to work."""
+    app = main.app
+
+    with app.test_request_context():
+        yield app
+
+
+@pytest.yield_fixture
+def firestore():
+
+    import firestore
+    """This fixture provides a modified version of the app's firebase model that
+    tracks all created items and deletes them at the end of the test.
+
+    Any tests that directly or indirectly interact with the database should use
+    this to ensure that resources are properly cleaned up.
+    """
+
+    # Ensure no books exist before running. This typically helps if tests
+    # somehow left the database in a bad state.
+    delete_all_books(firestore)
+
+    yield firestore
+
+    # Delete all books that we created during tests.
+    delete_all_books(firestore)
+
+
+# The backend data stores can sometimes be flaky. It's useful to retry this
+# a few times before giving up.
+@retry(
+    stop_max_attempt_number=3,
+    wait_exponential_multiplier=100,
+    wait_exponential_max=2000)
+def delete_all_books(firestore):
+    while True:
+        books, _ = firestore.next_page(limit=50)
+        if not books:
+            break
+        for book in books:
+            firestore.delete(book['id'])
+
+
+# Mark all test cases in this class as flaky, so that if errors occur they
```

Comment:
Nit: did you mean to keep this comment here?

Reply:
Nope! Removed.